Auto_ts Handler

pranavvp16 opened the pull request:
@martyna-mindsdb here is the PR and the issue we discussed on slack
<img width="432" alt="Screenshot 2023-10-09 at 8 11 00 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/f3974111-8f28-457a-baf1-edaf1fd0593a">
The above query gives error `Table not found for identifier: t.Time_Period` while trying to predict with the handler
#3499

martyna-mindsdb commented:
@pranavvp16 
Can you provide a csv file with sample data you are using please?

martyna-mindsdb commented:
@pranavvp16 
In the SELECT statement, replace the FROM clause to be like `FROM files.test_sales3 AS t`.

In my local env, it results in this error:

![image](https://github.com/mindsdb/mindsdb/assets/109554435/ebdeb84a-1071-4661-9afb-7d384e0022f2)

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/setup.py:
```diff
@@ -0,0 +1,21 @@
+from mindsdb.integrations.libs.const import HANDLER_TYPE
```
Contents in this file seem duplicated from `__init__.py`. Probably not the culprit, but would be good to fix this and follow the pattern that other ML handlers show.

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/requirements.txt:
```diff
@@ -0,0 +1,7 @@
+dask
```
Duplicate? If `dask[distributed]` is enough, then let's remove this.

paxcema reviewed (changes requested):
I don't think this integration is ready for review. The fitting process fails quite easily due to how arguments are being handled. Here's an example:

```Start of Fit.....
    Target variable given as = Marketing_Expense
Start of loading of data.....
    Inputs: ts_column = Time_Period, sep = ,, target = ['Marketing_Expense']
    Using given input: pandas dataframe...
    Time_Period column exists in given train data...
    train data shape = (18, 1)
Alert: Could not detect strf_time_format of Time_Period. Provide strf_time format during "setup" for better results.
setting maxlag to a low number = 5
Running Augmented Dickey-Fuller test with paramters:
    maxlag: 5 regression: c autolag: BIC
Data is already stationary
There is no differencing needed in this datasets for VAR model
    No time series plot since verbose = 0. Continuing
Time Interval is given as M
    Correct Time interval given as a valid Pandas date-range frequency...
==================================================
Building Auto SARIMAX Model
==================================================
Running Auto SARIMAX Model...
Exception occurred while building Auto SARIMAX model...
'>' not supported between instances of 'str' and 'int'
    Auto SARIMAX model error: predictions not available.
Best Model is: auto_SARIMAX
    Best Model (Mean CV) Score: inf
--------------------------------------------------
Total time taken: 0 seconds.
--------------------------------------------------
Leaderboard with best model on top of list:
            name  rmse
0  auto_SARIMAX   inf
```

Key part above being `Time Interval is given as M. Correct Time interval given as a valid Pandas date-range frequency...`. That means for this particular argument you need to interpret these strings and build the corresponding date range freq object from Pandas.

@pranavvp16 please use a debugger to verify the training process finishes correctly. Additionally, this will let you understand better where the error at prediction time is happening, exactly.

pranavvp16 commented:
```
Leaderboard with best model on top of list:
            name        rmse
0       Prophet   49.692177
3            ML   99.158912
1  auto_SARIMAX  114.708122
2           VAR  135.979141

```
I have now handled all the inputs correctly and the model train fine on all the types and shows the leaderboard of best model @paxcema

pranavvp16 commented:
<img width="756" alt="Screenshot 2023-10-10 at 8 31 57 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/81e8ca5d-13d7-46e7-843c-567be4399ab3">
just a last fix , Is there anything wrong in the SQL query

pranavvp16 commented:
<img width="380" alt="Screenshot 2023-10-11 at 11 37 56 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/26c95d61-cf65-410e-840f-d19b769086ac">
<img width="372" alt="Screenshot 2023-10-11 at 11 39 40 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/801408da-738b-4ef8-b778-c2a6455f6563">
<img width="804" alt="Screenshot 2023-10-11 at 11 40 10 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/57a061db-5ecb-46a7-9b9e-901c603b20e8">
<img width="879" alt="Screenshot 2023-10-11 at 11 40 19 PM" src="https://github.com/mindsdb/mindsdb/assets/94780581/0222a422-0be6-4c48-b8fb-4ead15284435">
Screenshots of handler working @paxcema

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/README.md:
```diff
@@ -0,0 +1,100 @@
+This integration is useful for time series data. It will automatically select the best model which gives best score specified. And it integrations with best Time Series 
+libraries like Statsmodels ARIMA, Seasonal ARIMA, Prophet and Scikit-Learn ML.
+
+The ideal use case is a very large dataset with many predictive features, where the user doesn't have strong priors about how these features may affect the target variable.
```
Is this true? Is the user to assume that exogenous variable support is first class, then?

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/README.md:
```diff
@@ -0,0 +1,100 @@
+
+### Why is this integration useful? What does the ideal predictive use case for this integration look like? When would you definitely not use this integration?
+This integration is useful for time series data. It will automatically select the best model which gives best score specified. And it integrations with best Time Series 
+libraries like Statsmodels ARIMA, Seasonal ARIMA, Prophet and Scikit-Learn ML.
```
`scikit-learn ML` should be renamed to `scikit learn`

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:
```diff
@@ -0,0 +1,116 @@
+        sep = ','
+        seasonal_period=None
+
+        if 'time_interval' in args:
```
This section seems overly verbose. I recommend the following pattern to set default values if the user does not provide custom values:

```python
param = args.get('param_name', default_value)
```

Doing this will let you do without lines 53-60

paxcema commented on tests/unit/executor_test_base.py:
```diff
@@ -120,6 +120,8 @@ def clear_db(self, db):
             name="langchain_embedding", data={}, engine="langchain_embedding"
         )
         db.session.add(r)
+        r = db.Integration(name="auto_ts", data={}, engine="auto_ts")
```
I don't see any unit tests. @ZoranPandovski should unit tests be created before accepting a submission? If so, please add some. Otherwise, please remove these two lines.

paxcema reviewed (changes requested):
Thanks for the changes @pranavvp16, looking good. While we wait for Zoran's answer regarding tests, there are still a couple unaddressed comments for when you've got time in order for us to move forward.

pranavvp16 commented:
@paxcema Fixed the pending ones directly from [docs](https://github.com/AutoViML/Auto_TS#tips), waiting for @ZoranPandovski

ZoranPandovski commented:
@pranavvp16, thanks for implementing all of the required changes. Let's add a few tests for create and create_validation. For reference, you can check https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_rag.py

pranavvp16 commented:
@ZoranPandovski I have added some unit tests related to create and create_validation

ZoranPandovski commented:
Please just pull the latest from staging so the OpenAI tests pass

pranavvp16 commented:
@ZoranPandovski done

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/README.md:
```diff
@@ -0,0 +1,99 @@
+USING
+    engine='auto_ts',
+    target = 'Sales',
+    ts_column = 'Time_Period',
```
Both `ts_column` and `target` columns should not be passed via `USING`, but through `ORDER BY` and `PREDICT` respectively. Please check this is the case, and also modify the README.

paxcema replied:
To be clear: examples below show usage of `PREDICT`, but the user should not need to repeat this name in `USING` later on.

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/README.md:
````diff
@@ -0,0 +1,99 @@
+``` sql
+SELECT m.* ,t.*
+FROM files.test_sales3 as t
+JOIN mindsdb.sales_model as m ;
````
This should use the `WHERE m.Time_Period > LATEST` syntax. Please refer to other time series handlers to conform to the same behavior.

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:
```diff
@@ -0,0 +1,96 @@
+        """
+        Creates Auto_ts model using the input df.
+        """
+        args = args['using']
```
This dict will contain a `timeseries_settings` key when created using time series syntax.

paxcema commented on mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:
```diff
@@ -0,0 +1,96 @@
+                    model_type=[model_type],
+                    verbose=0)
+
+        ts_column = args['ts_column']
```
Get from time series settings map instead of using.

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                FROM files
+                    (SELECT * FROM Sales_and_Marketing)
+                PREDICT Sales
+                USING
```
Change to time series specific model creation (i.e. `ORDER BY` clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                (SELECT * FROM Sales_and_Marketing)
+            PREDICT Sales
+            USING
+               engine="auto_ts"
```
Change to time series specific model creation (i.e. ORDER BY clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                (SELECT * FROM Sales_and_Marketing)
+            PREDICT Sales
+            USING
+               engine="auto_ts"
```
Change to time series specific model creation (i.e. ORDER BY clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                (SELECT * FROM Sales_and_Marketing)
+            PREDICT Sales
+            USING
+               engine="auto_ts"
```
Change to time series specific model creation (i.e. ORDER BY clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                (SELECT * FROM Sales_and_Marketing)
+            PREDICT Sales
+            USING
+               engine="auto_ts"
```
Change to time series specific model creation (i.e. ORDER BY clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+                (SELECT * FROM Sales_and_Marketing)
+            PREDICT Sales
+            USING
+               engine="auto_ts"
```
Change to time series specific model creation (i.e. ORDER BY clause)

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+            """
+                CREATE MODEL proj.auto_ts_missing_ts_column
+                FROM files
+                    (SELECT * FROM Sales_and_Marketing)
```
Do these tests run correctly on a fresh install? I don't see the `Sales_and_Marketing` file being added in the PR, so I suspect these may fail.

paxcema commented on tests/unit/ml_handlers/test_auto_ts.py:
```diff
@@ -0,0 +1,154 @@
+from ..executor_test_base import BaseExecutorTest
+
+
+class TestAuto_ts(BaseExecutorTest):
```
More generally, once you fix the points previously mentioned, please make sure tests actually call the predictor with `JOIN` and `LATEST`. Refer to our docs for examples on how this looks like. That will ensure this is a properly implemented time series integration (as in, it conforms to how others behave, too).

paxcema reviewed (dismissed):
Just noticed that there are some pending changes we require before being able to merge this. Thanks for the progress so far @pranavvp16, please take a look and address when you've got time.

pranavvp16 commented:
@paxcema I have added tests which check the predictions of the handler with `JOIN` and also fixed the way input of Time-Series Column was taken using `ORDER BY`. I have also the added the `sales.csv` in the data folder which will support tests

MinuraPunchihewa reviewed (commented):
Hey @pranavvp16,
I tried creating a model by running your branch locally, but I am getting the following error,
![image](https://github.com/mindsdb/mindsdb/assets/49385643/baf65fae-d352-417f-8708-43129e3f09f7)

Did you come across this?
@dusvyat Any ideas?

pranavvp16 commented:
@MinuraPunchihewa there's no such exception thrown from the create validation method, can you share the console output

MinuraPunchihewa commented:
@pranavvp16 This is it,
```
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/http/namespaces/sql.py", line 40, in post
    result = mysql_proxy.process_query(query)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/mysql_proxy.py", line 495, in process_query
    executor.query_execute(sql)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor.py", line 116, in query_execute
    self.do_execute()
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor.py", line 213, in do_execute
    ret = self.command_executor.execute_command(self.query)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py", line 579, in execute_command
    return self.answer_create_predictor(statement)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/functions.py", line 68, in wrapper
    return func(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py", line 1551, in answer_create_predictor
    df = self.session.model_controller.create_model(statement, ml_handler)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/interfaces/model/model_controller.py", line 315, in create_model
    predictor_record = ml_handler.learn(**params)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/integrations/libs/ml_exec_base.py", line 168, in learn
    self.handler_class.create_validation(
TypeError: create_validation() got an unexpected keyword argument 'handler_storage'
```
It seems to be coming from the core MindsDB code though, however, I am not experiencing this for the other handlers.

pranavvp16 commented:
@MinuraPunchihewa @paxcema can I make a loom video demonstrating the working of handler, cause it works on my device and also works on my ubuntu desktop

MinuraPunchihewa commented:
@pranavvp16 Yes, please do. That would help.

ea-rus commented on mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:
```diff
@@ -0,0 +1,93 @@
+    name = 'auto_ts'
+
+    @staticmethod
+    def create_validation(self, args):
```
As I look at other handlers code, interface of this function should be:
```python
def create_validation(target, args=None, **kwargs):
```
It started to receive new param handler_storage BTW

MinuraPunchihewa reviewed (commented):
Hey @pranavvp16,
It seems to work fine now, but there is one little concern that I have, your logic seems to add the suffix _preds to the column that predictions are made on. Is there a reason for this?
I'm just thinking that if I create a model with `PREDICT Sales`, for instance, I would expect to query my predictions using the `Sales` column.

pranavvp16 commented:
fine I'll change that anything else @MinuraPunchihewa

MinuraPunchihewa commented:
@pranavvp16 That is about it right now.

pranavvp16 commented:
so should I change the prediction column name to sales or keep it as it is ? @MinuraPunchihewa

MinuraPunchihewa commented:
@pranavvp16 It should be whatever comes after the `PREDICT` keyword.

pranavvp16 commented:
fine I'll make the change

dusvyat commented:
from my side, thanks for the contribution @pranavvp16 . After you make the suggested change, it is good to go

pranavvp16 commented:
@dusvyat welcome, @MinuraPunchihewa I have changed the prediction column accordingly. So are we ready for merge ?

MinuraPunchihewa commented:
@pranavvp16 @chandrevdw31 I believe we are ready to merge this.
@pranavvp16 Can you please check and resolve the CI errors that are popping up. I think you have a bunch of dependencies defined in your requirements.txt that are not in use.

By the way, one tiny thing I forgot, can you update the examples in your README with the last change that we did (by removing '_preds')?

pranavvp16 commented:
@MinuraPunchihewa thanks for the suggestion I have changed `README.md` and `requirements.txt` accordingly. @chandrevdw31 let's get this merged

chandrevdw31 commented:
@MinuraPunchihewa @pranavvp16 will merge as soon as checks are done

pranavvp16 commented:
@chandrevdw31 @MinuraPunchihewa  I think the requirements check fails due to version conflicts of auto-ts. What to do in such scenario

pranavvp16 commented:
@MinuraPunchihewa Looking at the failing tests what I infer is it tries to check whether the requirements match import but fails as requirements has `auto-ts` and import is `auto_ts`. If this is the issue it needs to be fixed in the CI. Please let me know if it isn't this and how can I fix it

pranavvp16 commented:
After a work around I found the issue is with `auto-ts`, I will make a PR in the repository and fix the issue

MinuraPunchihewa commented:
Hey @pranavvp16,
If you've made the PR, you can link it here. Just to keep a track of it.

pranavvp16 commented:
till then can we `pip install <my_fork>` which is a fixed version of auto_ts ?. I'll change it as soon it is fixed in the main auto_ts repo

MinuraPunchihewa commented:
@pranavvp16 I believe we have made the decision to avoid installing forks at the moment, unfortunately.
@StpMax Could you advice us on how best go about this?

StpMax commented:
@MinuraPunchihewa you mean issue with `auto-ts`->`auto_ts`? here is small PR to fix it: https://github.com/mindsdb/mindsdb/pull/8417

MinuraPunchihewa commented:
@StpMax Thanks, Max.
@pranavvp16 I see that your other PR has been merged. I guess we now need to wait for the new release.
By the way, there also seems to be a small issue that Flake has picked up. Can you please fix that?

pranavvp16 commented:
@MinuraPunchihewa auto-ts has released the latest version and I was also fixed the flake8 error

MinuraPunchihewa commented:
Hey @pranavvp16,
That's great! A couple of CI checks are still failing. Can you please have a look?
@paxcema If you have any other comments on this integration, they are certainly welcome!

MinuraPunchihewa commented:
Hey @pranavvp16,'
I believe the staging branch of your fork is behind staging branch of the original `mindsdb` repository. Can you please sync your fork and then merge staging into your branch?

MinuraPunchihewa reviewed (changes requested):
@pranavvp16 I'm really sorry for requesting you make changes again, but can we please do two things (I believe these will resolve the CI issues as well),
1. Let's remove the `setup.py` file; it is not required. After speaking to @paxcema, I understood that this was part of the old design for integrations and is not in practice currently.
2. Let's discard the changes that you've made to `executor_test_base.py`; after speaking to @ea-rus I understood that the line `r = db.Integration(name="auto_ts", data={}, engine="auto_ts")` is basically used is somewhat the equivalent for the `CREATE ML_ENGINE` command. If you discard your changes, it will fix the CI issue where there is no new line at the end of the file.
3. Let's improve your tests a little bit; there is a BaseMLAPITest class that you can inherit from to make your code a little more DRY. Some of the common functions like `wait_predictor()` and `run_sql()` are implemented here. Further, let's add a `setup_method()` function to your test class, so that you can put some of the common code in there including the `self.run_sql("create database proj")` line **AND** the `CREATE ML_ENGINE` command which we have removed in 2 above. You can refer the [tests for Anthropic](https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/base_ml_test.py) to get a better idea of this.

Again, apologies for asking for more changes after the PR was approved.

pranavvp16 commented:
The test you shared doesn't implement any of the methods instead it has methods like `run_sql` and `wait_predictor`. I was asked by Zoran to write the tests where he shared me this [test](https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_rag.py) example and asked to add create and validation tests only, later I also added predict tests too. Please share me a better example that implements the class your talking about @MinuraPunchihewa
